proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,314 @@
+::
+
+  error s"Expected: ${x + y}, got: ${result}"
+
+  let textExample = s"${name1} (age: ${age1}) encountered ${name2} (age: ${age2})"
+
+Proposed Change Specification
+-----------------------------
+
+When a new ``StringInterpolation`` extension is enabled, the following syntax:
+::
+
+  s"foo ${f a b} bar ${g x} baz ${name}"
+
+will be syntax sugar for:
+::
+
+  ("foo " ++)
+  . interpolatePrec 0 (f a b)
+  . (" bar " ++)
+  . interpolatePrec 0 (g x)
+  . (" baz " ++)
+  . interpolatePrec 0 name
+  $ ""
+
+where:
+::
+
+  class Interpolate a where
+    {-# MINIMAL interpolate | interpolatePrec #-}
+
+    interpolate :: a -> String
+    interpolate a = interpolatePrec 0 a ""
+
+    interpolatePrec :: Int -> a -> ShowS
+    interpolatePrec _ a s = interpolate a ++ s
+
+    -- personally I would rather leave this out and just use
+    -- OVERLAPPING + TypeSynonymInstances to define
+    -- `Interpolate String`, but I'll mimic Show entirely for now
+    interpolateList :: [a] -> ShowS
+    interpolateList xs = ('[':) . go True xs . (']':)
+      where
+        go _ [] = id
+        go isStart (x:xs) = (if isStart then id else (',' :)) . interpolatePrec 0 x . go False xs
+
+with instances mostly mirroring ``Show``, except for things like ``Char`` or ``String``, which would not include the quotes.
+
+This proposal adds a new class instead of ``Show`` because ``Show`` is intended to return the Haskell source code used to create the value. A quick search indicates that nothing on Stackage currently defines an ``Interpolate`` type class (alternatively, the class + functions could be named ``Display``, which is only in use by ``rio``).
+
+Could optionally provide a newtype for easy derivation with ``DerivingVia``, or even make ``Interpolate`` one of the built-in class that can be derived (like ``Show``):
+::
+
+  newtype InterpolateFromShow a = InterpolateFromShow a
+
+  instance Interpolate (InterpolateFromShow a) where
+    interpolatePrec = showsPrec
+
+  data Foo = Foo Int Bool
+    deriving (Show)
+    deriving (Interpolate) via InterpolateFromShow Foo
+
+We will always require interpolation to be of the form ``${foo}``, unlike other languages where ``$foo`` is allowed for single-variable interpolation. This simplifies the syntax and avoids the need for escaping bare ``$``. Interpolation may be escaped by escaping the dollar sign; e.g. ``"\${foo}"`` is equivalent to ``['$', '{', 'f', 'o', 'o', '}']``.
+
+Any Haskell expression may go inside the braces, but an unmatched ``}`` will immediately end processing. This restriction simplifies the implementation and isn't a big deal in practice, since one could always break it out into a helper variable.
+
+When ``OverloadedStrings`` is enabled, calls ``fromString`` with the result.
```

Comment:
This means that SQL libraries which currently provide safe `IsString` instances retroactively become vulnerable to injection, if someone tries to interpolate a query string.

Reply:
Yeah, that's fair. I added a note to that